### PR TITLE
Delivery process

### PR DESCRIPTION
**Pull request by gijskant**
- Delivery process definition
- Delivery process service and delivery service
- Delivery resource.

### CODE REVIEW

File: podium-common/src/main/java/nl/thehyve/podium/common/service/dto/DeliveryProcessRepresentation.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    private UUID uuid;
+
+    private DeliveryStatus status = DeliveryStatus.None;
+
+    private DeliveryProcessOutcome outcome = DeliveryProcessOutcome.None;
+
+    private RequestType type;
+
+    private String reference;
```
Comment (berndvdveen) [inline podium-common/src/main/java/nl/thehyve/podium/common/service/dto/DeliveryProcessRepresentation.java]: Is this a String or a DeliveryReferenceRepresentation?

Comment (gijskant) [inline podium-common/src/main/java/nl/thehyve/podium/common/service/dto/DeliveryProcessRepresentation.java]: String. This `DeliveryReferenceRepresentation` is purely a wrapper to be able to do POST/PUT requests.

File: podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryProcessService.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    public DeliveryProcess received(AuthenticatedUser user, DeliveryProcess deliveryProcess) throws ActionNotAllowedInStatus {
+        if (deliveryProcess.getStatus() == DeliveryStatus.Released) {
+            setVariable(deliveryProcess, DeliveryVariable.Received, Boolean.TRUE);
+            return completeCurrentTask(user, deliveryProcess);
+        }
+        throw ActionNotAllowedInStatus.forStatus(deliveryProcess.getStatus());
+    }
+
+    public DeliveryProcess cancel(AuthenticatedUser user, DeliveryProcess deliveryProcess) throws ActionNotAllowedInStatus {
+        if (deliveryProcess.getStatus() == DeliveryStatus.Released) {
```
Comment (berndvdveen) [inline podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryProcessService.java]: You should also be able to cancel a deliveryProcess when its not in 'Released'. This is to ensure maximum flexibility of the process flow and support cancellation of deliveries also when they're being 'Prepared'.

File: podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryService.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+        Optional<DeliveryProcess> deliveryProcessOptional = request.getDeliveryProcesses().stream()
+            .filter(process -> process.getUuid().equals(deliveryProcessUuid)).findAny();
+        if (deliveryProcessOptional.isPresent()) {
+            return deliveryProcessOptional.get();
+        }
+        throw new ResourceNotFound("Could not find delivery process with uuid " + deliveryProcessUuid.toString());
+    }
+
+    /**
+     *
```
Comment (berndvdveen) [inline podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryService.java]: Docs

File: podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryService.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+     * @throws ActionNotAllowedInStatus iff the request is not in status Approved.
+     */
+    public List<DeliveryProcessRepresentation> startDelivery(AuthenticatedUser user, UUID uuid) throws ActionNotAllowedInStatus {
+        Request request = requestRepository.findOneByUuid(uuid);
+        checkStatus(request, RequestStatus.Approved);
+        List<DeliveryProcessRepresentation> deliveryProcesses = new ArrayList<>();
+        for(RequestType type: request.getRequestDetail().getRequestType()) {
+            DeliveryProcess deliveryProcess = deliveryProcessService.start(user, type);
+            request.addDeliveryProcess(deliveryProcess);
+            publishDeliveryStatusUpdate(user, RequestStatus.Approved, request, deliveryProcess, null);
```
Comment (berndvdveen) [inline podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryService.java]: We could make the sourceStatus dynamic (for all events) using request.getStatus().

File: podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryService.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            deliveryProcesses.add(deliveryProcessMapper.deliveryProcessToDeliveryProcessRepresentation(deliveryProcess));
+        }
+        request.setStatus(RequestStatus.Delivery);
+        request = requestRepository.save(request);
+        requestService.publishStatusUpdate(user, RequestStatus.Approved, request, null);
+        return deliveryProcesses;
+    }
+
+    /**
+     *
```
Comment (berndvdveen) [inline podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryService.java]: Docs

File: podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryService.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+        deliveryProcess = deliveryProcessService.release(user, deliveryProcess);
+        deliveryProcess.setReference(reference.getReference());
+        requestRepository.save(request);
+        publishDeliveryStatusUpdate(user, DeliveryStatus.Preparation, request, deliveryProcess, null);
+        return deliveryProcessMapper.deliveryProcessToDeliveryProcessRepresentation(deliveryProcess);
+    }
+
+    /**
+     *
+     * @param user
```
Comment (berndvdveen) [inline podium-gateway/src/main/java/nl/thehyve/podium/service/DeliveryService.java]: Docs

File: podium-gateway/src/main/java/nl/thehyve/podium/service/RequestReviewProcessService.java [hunk 7]
```diff
@@ -161,23 +161,19 @@ private void setVariable(RequestReviewProcess requestReviewProcess, ReviewVariab
...
-    private static ActionNotAllowedInStatus actionNotAllowedInStatus(RequestReviewStatus status) throws ActionNotAllowedInStatus {
-        return new ActionNotAllowedInStatus("Invalid action in status: " + status.name());
-    }
-
     public RequestReviewProcess submitForValidation(AuthenticatedUser user, RequestReviewProcess requestReviewProcess) throws ActionNotAllowedInStatus {
         if (requestReviewProcess.getStatus() == RequestReviewStatus.Revision) {
             return completeCurrentTask(user, requestReviewProcess);
         }
-        throw actionNotAllowedInStatus(requestReviewProcess.getStatus());
+        throw ActionNotAllowedInStatus.forStatus(requestReviewProcess.getStatus());
```
Comment (berndvdveen) [inline podium-gateway/src/main/java/nl/thehyve/podium/service/RequestReviewProcessService.java]: Renaming the exception to ActionNotAllowed and then having the method .forStatus makes it nicer to read. I believe we fairly consistently use the .forStatus method already throughout the application.

File: podium-gateway/src/main/java/nl/thehyve/podium/web/rest/DeliveryResource.java [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+
+    private static final String ENTITY_NAME = "delivery";
+
+    @Autowired
+    private DeliveryService deliveryService;
+
+    @Autowired
+    private SecurityService securityService;
+
+
```
Comment (berndvdveen) [inline podium-gateway/src/main/java/nl/thehyve/podium/web/rest/DeliveryResource.java]: New line

**Pull request merged**